### ISSUE TITLE
Enhancement: [no-explicit-any] value-level only

### ISSUE THREAD

**Issue by geoffreytools** [typescript-eslint/typescript-eslint#9065]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-explicit-any/

### Description

I may want to prevent `any` from being used in a function body/signature while still allowing its use in type-level programming.

Not only is `any` unavoidable in some situations, or simply convenient, using it can also be an optimisation strategy, as internal type-checking of utility types can take a big share of compilation time in some projects.

### Fail

```TypeScript
function greet(): any {}
```

### Pass

```TypeScript
type UnionToIntersection<U> =
    (U extends any ? (k: U) => void : never) extends
        ((k: infer I) => void) ? I : never
```

**Comment by auvred** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/6437

Wouldn't replacing `any` with `U` in this particular example do the trick?

```diff
type UnionToIntersection<U> =
-   (U extends any ? (k: U) => void : never) extends
+   (U extends U ? (k: U) => void : never) extends
        ((k: infer I) => void) ? I : never
```

**Comment by auvred** [comment]
> `any` can appear anywhere in type-level programming, it is not harmful the way it is in value-level programming and anyone doing anything non-trivial tests their utility types anyway

I don't quite understand the difference between using `any` type at the "value level" and at the "type level". `any` is a type and can only be used at the type level, can't it?

If by "value-level" you mean, for example, the following:

```ts
function test(): any {}
```

Then how should we treat the following?

```ts
function test<T>(arg: T): T extends string ? number : any {}
```

This is a fairly simple conditional type, and using `any` here is obviously dangerous. But at the same time it is a conditional type, so it's kind of type-level programming. From my perspective,  it's almost impossible to distinguish the _**really**_ necessary use of `any` from any other use of `any` just by looking at the AST of the code.

---

I think 99% of the uses of `any` even in advanced type-level programming can be replaced by `unknown` or something else. And if you run into that 1% _**really**_ needed `any`, you can just put a `// eslint-disable` directive explaining why using `any` here is necessary.

If we decide to allow `any` in all the complex types, we give excessive freedom on the risky field of types to most users of this rule.

**Comment by Josh-Cena** [comment]
Agreed. While `any` is theoretically fine to use in type level operations, in order to make the option fool-proof we have to prove that the `any` does not leak into the resulting type and that's beyond our scope. Linting rules are at the end of the day to prevent dumbness—if you don't think there's value for a lint rule in a particular place because "you know better than a program" feel free to disable it.

**Comment by geoffreytools** [comment]
By "value-level" I meant anything that is directly assigned to a value, or a type cast.

I would not allow complex types using `any` to be written inline because such types could not be tested.

I realise that what I want is some kind of `skipLibCheck`. Realistically, any tested type would be imported because you tend not to test types in the file they are defined (testing is costly and verbose). Such types are really some kind of library.

---

Practically, the behaviour could be to stop looking for `any` at the first imported generic type that you encounter when exploring a type definition recursively from a value.

This should pass
```typescript
export type ToNumberOrAny<T> = T extends string ? number : any;
```
```typescript
import { ToNumberOrAny } from 'utils';

function test<T>(arg: T): ToNumberOrAny<T> {}
```

This should not
```typescript
type ToNumberOrAny<T> = T extends string ? number : any;

function test<T>(arg: T): ToNumberOrAny<T> {}
```

Nor this of course
```typescript
function test<T>(arg: T): T extends string ? number : any {}
```

---

It is not your responsibility to prove that `any` is leaking if the user decides to rely on testing to check for `any`. Using `any` explicitly is not the only way for a type to be inferred as `any` anyway. You can't win this battle.

At current if you work on a codebase with `no-explicit-any` enabled and you need to use some fancy type that uses `any`, you have to remove the lint rule altogether. It is certainly less safe than setting the discussed option.

---

I am not convinced that you can replace `any` with `unknown` because it would not work in situations where contra-variance is involved.

Moreover, I have a *hunch* that using `unknown` would not have the same performance implications than using `any` in a lot of contexts. Typescript is not only about type inference, it's also checking types inside your utility types and I doubt it checks `unknown` the same way it checks `any` in every code path.
